src/lib/Occurrence.hs

```diff
@@ -0,0 +1,623 @@
+--         2 -> compile opt2 ref
+--         ...
+--
+-- - `All parts` turns into sequentialization
+--     \ref.
+--       compile part1 ref
+--       compile part2 ref
+--       ...
+--
+-- - `ForEach b sub` is iteration
+--     \ref. for b in (type of b).
+--       compile sub ref
+
+-- === Use ===
+
+-- The Use structure is a "data space" representation of accessing
+-- arrays.
+--
+-- Note that a Use can still depend on iteration-space indices (from
+-- the `iter` scope)---it just doesn't have internal binders that are
+-- meant to be interpreted in iteration space.
+--
+-- A Use implicitly iterates over as many dimensions of the accessed
+-- object as there are nesting levels of At constructors in the Use.
+-- We do not use explicit binders for this because they always appear
+-- in the same nesting order and can always be pulled to the outermost
+-- layer, so it's simplest to leave them implicit.
+
+data Use (iter::S) =
+  -- The object is accessed in its entirety
+    Const Count
+  -- The object is accessed according to the third argument, but only
+  -- where the first argument equals the second.
+  | At (IxExpr iter) (Use iter)
+  | Max [Use iter]
+  | Sum [Use iter]
+  deriving (Eq, Show)
+
+-- The semantics of Use are similar to those for Access.  Here is the
+-- compilation that turns a Use into a program that dynamically counts
+-- how many times each index is used.
+--
+-- Note that now IxExpr permits the `IxAll` constructor, so an
+-- `IxExpr` represents a subset of indices rather than a single index.
+--
+-- - `Const ct` becomes accumulation
+--     \ref. ref += ct
+--     Also polymorphic in the ref type, applying elementwise
+--
+-- - `Max alts` becomes case
+--     \ref. case randint (length alts) of
+--       1 -> compile alt1 ref
+--       2 -> compile alt2 ref
+--       ...
+--
+-- - `Sum parts` becomes sequencing
+--     \ref.
+--       compile alt1 ref
+--       compile alt2 ref
+--       ...
+--
+-- - `At expr use` becomes testing and slicing
+--     \ref. for b in <next object dim>.
+--       if b in expr then (compile use ref!b)
+
+-- This semantics implies a transformation that we perform eagerly:
+-- - It's safe to distribute Sum into Max, so that Max ends up on the
+--   outside.
+
+-- All the At constructors created by `interp`, below, on any path
+-- through a sound Access refer to the same dimensions of the object
+-- being indexed, in the same order, with no gaps.  Why?  Because each
+-- is introduced at the same point in the sequence of index
+-- expressions applied to the object, regardless of the
+-- iteration-space control path that got there.
+--
+-- Another way to see this is that the compilation of At is the only
+-- thing that cares about and alters the type of the reference being
+-- passed around.  Ergo, all the At constructors must occur in the
+-- same order.
+
+-- === Use smart constructors ===
+
+-- We maintain a normal form for Use data with these smart
+-- constructors.  The main invariant is
```

Comment:
There are no smart constructors here, only the `MaxPlus` instance. But its methods don't seem sufficient to me? I think `Const One` is also a good normal form for a `Use`

Reply:
Outdated wording fixed.